src/prototree/base.py

```diff
@@ -0,0 +1,153 @@
+        self.backbone = backbone
+
+    def extract_features(self, x: torch.Tensor) -> torch.Tensor:
+        """
+        Applies the feature net and add_on layers to the input. The output
+        has the shape (batch_size, num_channels, height, width), where num_channels is the
+        number of channels of the prototypes.
+        """
+        x = self.backbone(x)
+        x = self.add_on(x)
+        return x
+
+    def patches(self, x: torch.Tensor) -> torch.Tensor:
+        """
+        Get the patches for a given input tensor. This is the same as extract_features, except the output is unfolded to
+        be (batch_size, d, n_patches_w, n_patches_h, w_proto, h_proto).
+        """
+        features = self.extract_features(x)
+        return self._features_to_patches(features)
+
+    def distances(self, x: torch.Tensor) -> torch.Tensor:
+        """
+        Computes the minimal distances between the prototypes and the input.
+        The output has the shape (batch_size, num_prototypes, n_patches_w, n_patches_h)
+        """
+        x = self.extract_features(x)
+        return self.proto_layer(x)
+
+    def patches_and_dists(self, x: torch.Tensor):
+        features = self.extract_features(x)
+        return self._features_to_patches(features), self.proto_layer(features)
+
+    def _features_to_patches(self, features: torch.Tensor) -> torch.Tensor:
+        w_proto, h_proto = self.prototype_shape[:2]
+        return features.unfold(2, w_proto, 1).unfold(3, h_proto, 1)
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        x = self.distances(x)
+        return torch.amin(x, dim=(2, 3))
+
+    @property
+    def num_prototypes(self):
+        return self.proto_layer.num_protos
+
+    @property
+    def prototype_channels(self):
+        return self.proto_layer.input_channels
+
+    @property
+    def prototype_shape(self):
+        return self.proto_layer.proto_shape
+
+    @torch.no_grad()
+    def project_prototypes(
+        self, node_to_patch_matches: dict[int, ImageProtoSimilarity]
+    ):
+        """
+        Replaces each prototype with a given patch.
+        Note: This mutates the prototype tensors.
+        TODO: We should probably not be mutating the tree (via the prototypes) after training, as this is making the
+         code less flexible and harder to reason about.
+        """
+        for proto_id, patch_info in node_to_patch_matches.items():
+            self.proto_layer.protos.data[proto_id] = patch_info.closest_patch.data
+
+    @torch.no_grad()
+    def update_proto_patch_matches(
+        self,
+        updated_matches: dict[int, ImageProtoSimilarity],
+        x: torch.Tensor,
+        y: torch.Tensor,
+    ):
+        """
+        Produces a map where each key is a node and the corresponding value is information about the patch (out of all
+        images in the dataset) that is most similar to node's prototype.
+
+        :param updated_matches:
```

Comment:
Could you add some info on what is expected in this parameter? How is it different from `node_to_patch_matches`?

Since this is a public method, it would be nice to add some info on how this dict is usually obtained. Same for other methods using non-trivial data structures, like `project_prototypes`

Reply:
Sorry, I realized that I forgot to update this part of the documentation and variable naming after refactoring it from the earlier code; I've made some changes now.